New libvirt driver for Cloud Hypervisor

linericyang opened the pull request:
The basic CH VM lifecycle management in libvirt has been implemented in this patch, include define/undefine/start/shutdown/destroy/suspend/resume. A domain definition xml show as below.
```
<domain type='ch'>
  <name>MyGuest</name>
  <uuid>4dea22b3-1d52-d8f3-2516-782e98ab3fa0</uuid>
  <genid>43dc0cf8-809b-4adb-9bea-a9abb5f3d90e</genid>
  <title>A short description - title - of the domain</title>
  <description>Some human readable description</description>
  <os>
    <type>hvm</type>
    <kernel>/root/cloud-hypervisor/hypervisor-fw</kernel>
  </os>
  <memory unit='G'>2</memory>
  <devices>
    <disk type='file'>
      <source file='/root/cloud-hypervisor/clear-31890-cloudguest.img'/>
      <target dev='hda'/>
    </disk>
  </devices>
  <vcpu>2</vcpu>
</domain>
```
 
It contains some debug code, please ignore them. Comments are welcome.

linericyang committed:
New driver for Cloud Hypervisor

sameo commented:
cc @sboeuf @rbradford @likebreath

sboeuf commented:
@linericyang that's quite nice, thanks for the PR :)
Do you have any pointers/documentation on how we can test this?

@rbradford @sameo how are we going to make sure this is tested from our CI?

rbradford commented:
@linericyang Great thank you! Would you mind describing how you use/test this assuming no libvirt knowledge?

linericyang commented:
1. At first build libvirt binary from source code. Assuming libvirt code has been checked out to path **<some_path>/libvirt**, and a build directory **<some_path>/build** has been created 
```
# manually install library dependencies, use ubuntu as example here
$ sudo apt install libyajl-dev libcurl4-openssl-dev

# build libvirt binary
$ cd <some_path>/build
$ ../libvirt/autogen.sh --system
$ make
```
Refer to [here](https://libvirt.org/compiling.html#building) for more compiling details.

2. Start libvirt daemon
```
$ <some_path>/build/src/libvirtd
```
3. Try every libvirt action to manage the CH CM, then issue "vm.info" to CH API socket to make sure VM state we got is expected. BTW, all CH socket are located in /var/run/libvirt/ch.
A typical workflow show as below.
```
# define a VM. You can use example domain xml in my previous message. It only store VM definition in internal structure, not actually start CH process at this point.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" define test.xml

# start VM. After execute this, VM should be booted successfully. Its status should be "running" in "vm.info" response.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" start MyGuest

# reboot VM. Its status should still be  "running" in "vm.info" response.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" reboot MyGuest

# suspend VM. Its status should be changed to "paused" in "vm.info" response.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" suspend MyGuest

# resume VM. Its status should be changed back to "running" in "vm.info" response.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" resume MyGuest

# shutdown VM. Its status should be changed to "created" in "vm.info" response.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" shutdown MyGuest

# destroy VM. The CH process should be killed by this command.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" destroy MyGuest

# undefine VM. Remove VM definition.
$ <some_path>/build/run ./tools/virsh -c "ch:///system" undefine MyGuest
```
Please note that currunt only part of CH parameters of VM creation have been supported in this patch. I am still working on this to achieve fully coverage. So please use above domain xml if you want to try it now. Let me know if you have any problem in this process.

sboeuf commented on src/ch/ch_conf.h:
```diff
@@ -0,0 +1,87 @@
+#include "virthread.h"
+
+#define CH_DRIVER_NAME "CH"
+#define CH_CMD "cloud-hypervisor"
```
This assumes there's a cloud-hypervisor binary on the system. What if we need to provide a specific path? Do we need to rebuild libvirt, or is there a more generic way of doing this?

linericyang replied:
Thank you so much for pointing out this. Typically user uses <emulator> element to specify the full path to binary as following shows. A new commit has been uploaded.
```
<devices>
  <emulator>/usr/bin/cloud-hypervisor</emulator>
</devices>
```

sboeuf reviewed (commented):
I didn't review the code in details (too many functions I don't know about), but I have one question to make the cloud-hypervisor path being changed more dynamically without having to rebuild libvirt.

sboeuf commented:
Thanks @linericyang, I really appreciate that we can start using CH with libvirt :)

I have tried out this PR, and overall it works fine.
I had to install some extra packages on my Ubuntu 20.04 machine in order to pass the configure step:
```
sudo apt install autoconf libtool libgnutls28-dev libnl-3-dev libnl-route-3-dev libxml2-dev libxml2-utils xsltproc docutils-common libdevmapper-dev libpciaccess-dev
```
I followed all the instructions and relied on the `test.xml` (adapted for my machine) to create/start/reboot/shutdown the VM, and I can say it works as expected.

We're missing some important devices like `net` or `console`, which would give the ability to access the guest and perform some checks that it's running as expected. The support for `net` would be the bare minimum to be able to automate these checks.
@linericyang what's your plan going forward with this?

@rbradford @sameo what should be the criteria to decide we can merge the first PR (given that more devices and features must be enabled through follow up PRs)?

linericyang committed:
Allow user to specify network interface

linericyang committed:
Support "emulator" element in domain xml

Allow user use "emulator" element to specify full qualified path to
cloud hypervisor binary.

linericyang commented:
Glad to know it works in your environment.

I created a new commit to support user to specify network interface. An example of net interface definition shows as below. 
```
<devices>
  <interface type='ethernet'>
    <target dev='mytap1'/>
    <mac address="00:11:22:33:44:55"/>
    <ip family='ipv4' address='172.17.2.0' prefix='24'/>
    <model type='virtio'/>
    <driver iommu='on' queues='4' rx_queue_size='256' tx_queue_size='256'/>
  </interface>
</devices>
```

It will generate following arguments in HTTP payload when issue PUT to /api/v1/vm.create.
```
"net": [{"ip":"172.17.2.0","mask":"255.255.255.0","tap":"mytap1","mac":"00:11:22:33:44:55","iommu":true,"num_queues":4,"queue_size":256}]
```

sboeuf commented:
@linericyang thanks a lot, I'll have a try!
It'd be good to rework the PR so that we can clearly identify the steps you went through to implement the CH driver here.
First commit would introduce the CH driver, and then multiple following commits would each introduce the support for a specific CH feature, like block, cpu, memory, ...
Once we have the PR ready like this, we can do one last review and decide to merge it.
Then, with follow up PRs, you can add each missing device/feature.
In the meantime, I'll work on creating some CI that hooks into the Jenkins from CH.

sboeuf commented:
Any progress on this @linericyang ?

linericyang commented:
@sboeuf Because of the project priority change, I will not continue this item anymore. But I heard another teams or colleagues will take over this to continue.

sboeuf commented:
@linericyang oh I'm sorry to hear that. Could you introduce these people here (github nicknames)? Just want to make sure we can keep making progress on this topic.

bryteise commented:
> It'd be good to rework the PR so that we can clearly identify the steps you went through to implement the CH driver here.
> First commit would introduce the CH driver, and then multiple following commits would each introduce the support for a specific CH feature, like block, cpu, memory, ...
> Once we have the PR ready like this, we can do one last review and decide to merge it.
> Then, with follow up PRs, you can add each missing device/feature.
> In the meantime, I'll work on creating some CI that hooks into the Jenkins from CH.

@sboeuf I'm going to start working on addressing your feedback here. Currently I'm just getting it back into working order after rebasing on the latest upstream which changed the build system over to meson.

sboeuf commented:
@bryteise Excellent! Let me know if you need anything :)

bryteise commented:
@sboeuf Would you be able to update the master branch to the latest upstream? That way the meson conversion changes are pulled in and my PR isn't going to be a bunch of upstream changes my patches are on top of.

sameo commented:
@bryteise The master branch is updated now.

bryteise commented:
@sameo Thanks!

sameo commented:
This PR has been replaced by PR #2
